### ISSUE TITLE
Hub stays offline, PetHubLocal closes TCP after TLS "Client Hello"

### ISSUE THREAD

**Issue by jacotec** [PetHubLocal/pethublocal#3]
After the last server desaster at Sure Petcare yesterday and today I've deployed PetHubLocal as HA addon (thanks @plambrechtsen 
 for mentioning it in the Twitter thread ;-) ) and everything looks pretty fine - except that my Hub stays offline and does not connect:

![image](https://user-images.githubusercontent.com/3321655/181272182-1e45295c-bfb7-4341-bf5c-9eb996be3ccf.png)

Tracing this with Wireshark the Hub (10.30.1.21) opens the TCP connection to the HA server (10.1.1.11) running the PHL. Then the Hub sends a "Client Hello" in TLS1.2 - the reaction of the PHL is to close the TCP connection right away.

![image](https://user-images.githubusercontent.com/3321655/181274110-4fce901a-2f0b-42ca-92d9-d3b4ab13cbdf.png)

No traces in puthublocal.log why this happens.

Hub runs the correct firmware:

![image](https://user-images.githubusercontent.com/3321655/181273649-53f6b16e-0294-4cfe-9955-859f263e78bb.png)

Any ideas how to get that up and running?

**Comment by jacotec** [comment]
Update: This issue is limited to the HA app only. It works with an Ubuntu VM with manual installation!

**Comment by plambrechtsen** [comment]
This will be because you are running `4.201` firmware and it checks the server certificate. You will need to flash with the `2.43` firmware automatically built using the XOR key found in the `4.201` firmware.

https://pethublocal.github.io/firmware

In short... boot your hub while holding down the reset button underneath and it will flash `2.43` on. Then if you want to roll back move the `2.43` firmware files out of the folder leaving only the bootloader version and reflash on the original firmware

**Comment by jacotec** [comment]
@plambrechtsen I don't think you. The config file says 2.43 as you can see above, also it works with manual installation (which it would not in that case). It's only in the HA app version.

Seems my hub never got the 2.201

**Comment by plambrechtsen** [comment]
Interesting. Sorry I didn't fully read the above and see yes you are running `2.43` in theory. 
When I saw that error it was absolutely because it was running the new firmware.
If you download the firmware to check.

`pethublocal firmware -sn Hxx`

That will download your current firmware from surepet. Then use `xxd Hxx...-00.bin | head 10`

It should give you the first firmware file header. In there the first two bytes are the CRC value. If you can post that I would be interested to know what it is.

**Comment by plambrechtsen** [comment]
Alternatively if you are happy to open the hub and solder on a TTL serial cable onto the console port I would be interested to know what comes via the console.

https://pethublocal.github.io/certificate

**Comment by jacotec** [comment]
Hi Peter,

here is the header of the firmware file -00.bin:

![image](https://user-images.githubusercontent.com/3321655/181441670-2915eed7-ce5d-42c9-be4c-1b5a9be96d55.png)

I tried to make the firmware update/reflash to be on the safe side, but when starting the hub with the reset button pressed, it just stays with continuously red ears. After nothing has happened for 15 minutes and it was still in that state, I just restarted it.

Meanwhile I'm not pretty sure about the animal status. Yesterday around midnight our Shadow changed to "outside" although she was definitely inside. Also I don't see the hub ears flashing green when a cat passes the flap ... so I have no indication whether an update was really sent. Is the update missing in this case?

**Comment by plambrechtsen** [comment]
Flashing ears is something that the cloud service sends to the hub. I personally found it annoying so didn't add it as a feature.

**Comment by jacotec** [comment]
Thinking further on the HA app issue ... if the hub would run a firmware checking the cert, the hub side would be the one closing the connection after getting a reply from the PHL.

As Wireshark shows the PHL closes the connection after the client hello from the hub, it appears as if the HA app version does not like TLSv1.2 due to whatever reason. The manual installation does not have this issue (I did not try the "normal" non-HA docker build, why having additional ballast if I need to run a separate VM at all)

**Comment by plambrechtsen** [comment]
Are you sure you don't have something else listening on port 443 @jacotec as I think I should add in some logic to exit if it can't bind on that port. If you do a `curl -k https://hostname/api/credentials' you should get a `405: Method Not Allowed` and see something like this in the logs:
```
INFO:aiohttp.access:127.0.0.1 "GET /api/credentials HTTP/1.1" 405 203 "-" "PostmanRuntime/7.29.0"
```

**Comment by jacotec** [comment]
@plambrechtsen Pretty sure, when I stop the app I see no listener on 443, also TCP will not connect. I can recheck the CURL tomorrow.

**Comment by jacotec** [comment]
@plambrechtsen Sorry for the delay, I've checked that now (need to be a bit careful to not connect to my main broker, that would cause issues with my production installation of PHL)

- Port 443 *is* exclusively used by PHL on the homeassistant VM. When the PHL add-on is stopped, noone can connect to 443.
- The CURL gives the 405 when I run PHL as app on the HA-VM

```
root@modbus2mqtt:~# curl -k https://10.1.1.11/api/credentials
curl: (7) Failed to connect to 10.1.1.11 port 443: Connection refused

*** started the app here ***

root@modbus2mqtt:~# curl -k https://10.1.1.11/api/credentials
405: Method Not Allowedroot@modbus2mqtt:~# 
```

So: Nope, there is definitely nothing else using port 443.

**Comment by flyize** [comment]
@jacotec Did you make any headway with this? Although I don't have Wireshark, I think I'm experiencing the same issue as you. Manual install works, but the HA Addon does not.

**Comment by it-guy92** [comment]
I don't have Wireshark, but i think i have the same issue.

**Comment by FeLLa-23** [comment]
Hello, I also have the same issue. Working perfectly fine with manual installation but no chance over HA addon integration.

**Comment by circa1665** [comment]
Have the same issue.

**Comment by circa1665** [comment]
Is there any update on this?

**Comment by plambrechtsen** [comment]
Yeah, SureFlap have updated the firmware and locked everyone out. If you have newer firmware than `2.43` such as `233..` then it's all busted unfortunately. Time to see if anyone wants to help me revive WemosHub.

**Comment by flyize** [comment]
While I know some python I can only assume that I don't have the skills to be of any real assistance. If it would help, I'd be happy to donate money to the cause though.

**Comment by tinuva** [comment]
How do I know what version my hub is on?
```
$ xxd H004-0088731-1.177-00.bin | head -n 100
00000000: 3465 3062 2031 6430 3030 3030 3020 3030  4e0b 1d000000 00
00000010: 3630 2031 6430 3030 3030 3020 3130 3030  60 1d000000 1000
00000020: 2030 3120 27df 0b76 077d a012 8c1c 9e82   01 '..v.}......
00000030: 536b 0b3c d720 00fe 777d aa12 fc1c 9482  Sk.<. ..w}......
```

**Comment by plambrechtsen** [comment]
Closing this issue as I now have a whole new approach. This python code is great for a reference but I won't be making any changes to it going forward. If you are interested in my new approach send me an email pethublocal at gmail . com.
